drivers/platform/xilinx/xilinx_spi_pl.c

```diff
@@ -0,0 +1,301 @@
+	const struct spi_init_param *param)
+{
+	struct spi_desc		*ldesc;
+	struct xspi_desc	*xdesc;
+	int32_t			err;
+	XSpi_Config		*cfg;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	err = -EINVAL;
+	cfg = XSpi_LookupConfig(param->device_id);
+	if (!cfg)
+		goto error;
+
+	if (param->chip_select > cfg->NumSlaveBits)
+		return -EINVAL;
+
+	err = -ENOMEM;
+	ldesc = (struct spi_desc *)calloc(1, sizeof(*ldesc));
+	if (!ldesc)
+		goto error;
+
+	xdesc = (struct xspi_desc *)calloc(1, sizeof(*xdesc));
+	if (!xdesc)
+		goto error;
+
+	xdesc->base_addr = cfg->BaseAddress;
+	xdesc->disable_slaves_mask = (1 << cfg->NumSlaveBits) - 1;
+
+	err = _xil_spi_init(xdesc);
+	if (IS_ERR_VALUE(err))
+		goto error;
+
+	ldesc->extra = xdesc;
+	*desc = ldesc;
+
+	return SUCCESS;
+
+error:
+	free(xdesc);
+	free(ldesc);
+
+	return err;
+}
+
+static int32_t xil_spi_remove_pl(struct spi_desc *desc)
+{
+	if (!desc)
+		return -EINVAL;
+
+	free(desc->extra);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/* Fill fifo until fifo is full or no more data available */
+static inline void _wr_fifo(struct xspi_desc *xdesc, uint8_t *data,
+		     uint32_t *remaining)
+{
+	uint32_t i;
+
+	i = _read(xdesc, XSP_TFO_OFFSET);
+	while (i < FIFO_DEPTH && (*remaining)) {
+		_write(xdesc, XSP_DTR_OFFSET, data[i]);
+		(*remaining)--;
+		i++;
+	}
+}
+
+/* Read fifo until fifo is empty or no more buffer available to read into */
+static inline void _rd_fifo(struct xspi_desc *xdesc, uint8_t *data,
+		     uint32_t *remaining)
+{
+	uint32_t i;
+	uint32_t rx_available;
+
+	rx_available = _read(xdesc, XSP_RFO_OFFSET);
+	i = 0;
+	while (i <= rx_available && *remaining) {
+		data[i] = _read(xdesc, XSP_DRR_OFFSET);
+		(*remaining)--;
+		i++;
+	}
+}
+
+/* If assert is 0 deassert CS, else assert */
+static int32_t xil_spi_cs_assert(struct spi_desc *desc, uint8_t assert)
+{
+	struct xspi_desc	*xdesc;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	xdesc = desc->extra;
+	if (assert)
+		_write(xdesc, XSP_SSR_OFFSET, ~(1 << desc->chip_select));
+
+	else
+		_write(xdesc, XSP_SSR_OFFSET, xdesc->disable_slaves_mask);
+
+	return SUCCESS;
+}
+
+/* If start is 0 stop stransfer, else start it */
+static inline void _xil_spi_start_transfer(struct xspi_desc *xdesc, uint8_t start)
+{
+	_update(xdesc, XSP_CR_OFFSET, start ? 0 : XSP_CR_TRANS_INHIBIT_MASK,
+		XSP_CR_TRANS_INHIBIT_MASK);
+}
+
+static int32_t xil_spi_write_and_read_pl(struct spi_desc *desc, uint8_t *data,
+				uint16_t bytes_number)
+{
+	struct xspi_desc	*xdesc;
+	uint32_t		tx_remaining;
+	uint32_t		rx_remaining;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	xdesc = desc->extra;
+	_update_mode(desc);
+
+	xil_spi_cs_assert(desc, 1);
+	_xil_spi_start_transfer(xdesc, 1);
+
+	rx_remaining = bytes_number;
+	tx_remaining = bytes_number;
+	while (tx_remaining || rx_remaining) {
+		_wr_fifo(xdesc, data + (bytes_number - tx_remaining),
+				&tx_remaining);
+		_rd_fifo(xdesc, data + (bytes_number - rx_remaining),
```

Comment:
Are you reading data before a `_xil_spi_start_transfer` call?

Reply:
_xil_spi_start_transfer(xdesc, 1) will start the transfer
_xil_spi_start_transfer(xdesc, 0) will stop the transfer
This code is between them.

Reply:
True